# PMB-project
The aim of this project is the identification of amino acids starting from the genome sequence.  
This would serve to calculate the amino acid requirements in different populations and see if and how this correlates with diet.

### 1000 Genome Project
The human genomes considered in this study are from the **1000 Genomes Project** (https://www.internationalgenome.org/data/
).  
This latter consist in a catalogue of common human genetic variation. It has been developed using openly consented samples from people who declared themselves to be healthy at the time of collection.  
It was the first project to sequence the genomes of a large number of people and data was quickly made available to the worldwide scientific community through freely accessible public databases.  
Infact, the reference data resources generated by the project are heavily used by the biomedical science community.  
The final data set (associated with the third and final phase: 2nd May 2013) contains data for 2,504 individuals from 26 populations and 84.4 million variants. 
Low coverage and exome sequence data are present for all of these individuals, 24 individuals were also sequenced to high coverage for validation purposes.  

### Pygeno package
The package that has been chosen in order to conduct this project analysis is called **Pygeno** (http://pygeno.iric.ca/index.html), a Phyton package for precision medicine and proteogenomics. 

![Pygeno Logo](logo.png)

### VCF files
The data that was analyzed was provided in **VCF format** (Variant Call Format).  
Is a format used in bioinformatics for storing gene sequence variations.
In this format only the variations are stored with a reference genome. 
The structure of the file is divided into an header, which provides metadata describing the body of the file, and a body, which contains the actual data. 

It is important to keep in mind that VCF files come in many formats an Pygeno only supports a certain number of them.   
In our case some problem with the parser occurred and files generated from PLINK 
have been chosen as they came already filtered, so as not to modify the Pygeno parser.

### titolo?????? DA QUI SISTEMARE
To estimate from the genomic data, in the VCF format, the amino acidic requirement.
